src/models/pretraining.py

```diff
@@ -0,0 +1,119 @@
+from src.models.vtr.ocr import OCRHead
+
+
+class Pretrain(nn.Module):
+    def __init__(
+        self,
+        emb_size: int = 512,
+        n_head: int = 8,
+        n_layers: int = 4,
+        device: str = "cpu",
+        ocr: OCRHead = None,
+        char2int_dict: dict = None,
+        alpha: float = 1,
+    ):
+        super().__init__()
+        config = BertConfig(hidden_size=emb_size, num_attention_heads=n_head)
+        self.encoder = BertModel(config)
+        self.positional_enc = PositionalEncoding(emb_size)
+        self.positional_dec = PositionalEncoding(emb_size)
+        self.masking = SpanMaskingGenerator(emb_size)
+        self.linear = nn.Linear(emb_size, emb_size)
+        self.decoder = nn.TransformerEncoder(
+            nn.TransformerEncoderLayer(d_model=emb_size, nhead=n_head, dim_feedforward=emb_size), num_layers=n_layers
+        )
+        self.ocr = ocr
+        self.ctc_criterion = CTCLoss(reduction="sum", zero_infinity=True)
+        self.char2int_dict = char2int_dict
+        self.device = device
+        #self.criterion = lpips.LPIPS(net="vgg").to(device)
+        self.criterion = nn.MSELoss()
+        self.emb_size = emb_size
+        self.alpha = alpha
+        self.iter = 0
+
+    def forward(self, input_batch: dict[str, list | torch.Tensor]):
+
+        batch_size, slice_count, height, width = input_batch["slices"].shape
+        slices = input_batch["slices"].view(batch_size, slice_count, height * width)
+        pos_slices = self.positional_enc(slices)
+        slices_detached = pos_slices.detach()
+
+        grey_slice = torch.full((height, width), 128, dtype=torch.float32, device=self.device).flatten()
+        masks = []
+        unmasked_slices = []
+        unmasked_texts = []
+        for i in range(batch_size):
+            mask = self.masking(len(slices[i]))
+            masks.append(mask)
+            masked_idx = np.where(mask == 1)[0]
+            masked_idx = masked_idx[masked_idx < len(slices_detached[i])]
+            if masked_idx.size < 10:
+                masked_idx = np.arange(10)
+            slices_detached[i][masked_idx] = grey_slice
+            if self.ocr:
+                unmasked_idx = np.where(mask == 0)[0]
+                unmasked_slices.append(slices[i][unmasked_idx[: len(slices[i]) - 1]])
+                unmasked_texts.append(
+                    np.array(input_batch["texts"][i])[unmasked_idx[: len(input_batch["texts"][i]) - 1]]
+                )
+
+        slices = self.linear(slices_detached)
+        encoded_text = self.encoder(inputs_embeds=slices)
+
+        encoded_text = self.positional_dec(encoded_text[0])
+        decoded_text = self.decoder(encoded_text)
+
+        masked_slices = []
+        masked_originals = []
+        for i in range(batch_size):
+            masked_idx = np.where(masks[i] == 1)[0]
+            masked_idx = masked_idx[masked_idx < len(slices_detached[i])]
+            if masked_idx.size < 10:
+                masked_idx = np.arange(10)
```

Comment:
I think that it's worth to save the masked_idx to a variable and don't copy-paste these 3 lines